ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidatorSpec.scala

```diff
@@ -0,0 +1,393 @@
+      val validator = BatchedSubmissionValidator[Unit](
+        BatchedSubmissionValidatorParameters.default,
+        engine,
+        metrics)
+      val batchSubmission = DamlSubmissionBatch.newBuilder
+        .addSubmissions(
+          CorrelatedSubmission.newBuilder
+            .setCorrelationId(aCorrelationId)
+            .setSubmission(anInvalidEnvelope))
+        .build
+
+      validator
+        .validateAndCommit(
+          Envelope.enclose(batchSubmission),
+          aCorrelationId,
+          newRecordTime().toInstant,
+          aParticipantId,
+          mock[DamlLedgerStateReader],
+          mock[CommitStrategy[Unit]]
+        )
+        .failed
+        .map { result =>
+          result shouldBe a[ValidationFailed]
+        }
+    }
+
+    "validate a non-batched submission" in {
+      val mockLedgerStateReader = mock[DamlLedgerStateReader]
+      val mockCommit = mock[CommitStrategy[Unit]]
+      val partySubmission = makePartySubmission("foo")
+      // Expect two keys, i.e., to retrieve the party and submission dedup values.
+      when(mockLedgerStateReader.readState(argThat((keys: Seq[DamlStateKey]) => keys.size == 2)))
+        .thenReturn(Future.successful(Seq(None, None)))
+      val logEntryCaptor = ArgumentCaptor.forClass(classOf[DamlLogEntry])
+      val outputStateCaptor = ArgumentCaptor.forClass(classOf[Map[DamlStateKey, DamlStateValue]])
+      when(
+        mockCommit.commit(
+          any[ParticipantId],
+          any[String],
+          any[DamlLogEntryId],
+          logEntryCaptor.capture(),
+          any[Map[DamlStateKey, Option[DamlStateValue]]],
+          outputStateCaptor.capture()))
+        .thenReturn(Future.unit)
+      val validator = BatchedSubmissionValidator[Unit](
+        BatchedSubmissionValidatorParameters.default,
+        engine,
+        metrics)
+
+      validator
+        .validateAndCommit(
+          Envelope.enclose(partySubmission),
+          aCorrelationId,
+          newRecordTime().toInstant,
+          aParticipantId,
+          mockLedgerStateReader,
+          mockCommit
+        )
+        .map { _ =>
+          // Verify that the log entry is committed.
+          logEntryCaptor.getAllValues should have size 1
+          val logEntry = logEntryCaptor.getValue.asInstanceOf[DamlLogEntry]
+          logEntry.getPartyAllocationEntry should be(partySubmission.getPartyAllocationEntry)
+
+          // Verify that output state contains the expected values.
+          outputStateCaptor.getAllValues should have size 1
+          val outputState =
+            outputStateCaptor.getValue.asInstanceOf[Map[DamlStateKey, DamlStateValue]]
+          outputState should have size 2
+          outputState.keySet should be(partySubmission.getInputDamlStateList.asScala.toSet)
+        }
+    }
+
+    "validate a batch" in {
+      validateBatchSubmission(nSubmissions = 1000, commitParallelism = 2)
+    }
+
+    "validate a batch with 1 committer for each submission" in {
```

Comment:
I see diminishing returns with more tests here but let me know if you have specific suggestions.

Reply:
I was mostly thinking about less submissions than committers.

Reply:
In that case we won't be using some of the committer threads. We would be testing akka instead of our logic. The other test cases make sure certain edge cases work that we need to rely on.